tracing.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018-2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"context"
+	"io"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/uber/jaeger-client-go/config"
+)
+
+const (
+	jaegerAgentHost = "127.0.0.1"
+
+	// This is the default.
+	jaegerAgentPort = "6831"
+)
+
+// Implements jaeger-client-go.Logger interface
+type traceLogger struct {
+}
+
+// tracerCloser contains a copy of the closer returned by createTracer() which
+// is used by stopTracing().
+var tracerCloser io.Closer
+
+func (t traceLogger) Error(msg string) {
+	agentLog.Error(msg)
+}
+
+func (t traceLogger) Infof(msg string, args ...interface{}) {
+	agentLog.Infof(msg, args...)
+}
+
+func createTracer(name string) (opentracing.Tracer, error) {
+	cfg := &config.Configuration{
+		ServiceName: name,
+
+		// If tracing is disabled, use a NOP trace implementation
+		Disabled: !tracing,
+
+		// Note that span logging reporter option cannot be enabled as
+		// it pollutes the output stream which causes (atleast) the
+		// "state" command to fail under Docker.
+		Sampler: &config.SamplerConfig{
+			Type:  "const",
+			Param: 1,
+		},
+
+		Reporter: &config.ReporterConfig{
+			// Specify the default values since without them,
+			// Jaeger will attempt to call the DNS resolver and
+			// that will fail since the agent runs relatively
+			// early in the boot sequence!
+			LocalAgentHostPort: jaegerAgentHost + ":" + jaegerAgentPort,
+
+			// FIXME: debugging only!
+			LogSpans: true,
+		},
+	}
+
+	logger := traceLogger{}
+
+	tracer, closer, err := cfg.NewTracer(config.Logger(logger))
+	if err != nil {
+		return nil, err
+	}
+
+	// save for stopTracing()'s exclusive use
+	tracerCloser = closer
+
+	// Seems to be essential to ensure non-root spans are logged
+	opentracing.SetGlobalTracer(tracer)
+
+	return tracer, nil
+}
+
+// stopTracing() ends all tracing, reporting the spans to the collector.
+func stopTracing(ctx context.Context) {
+	if !tracing {
+		return
+	}
+
+	span := opentracing.SpanFromContext(ctx)
+	if span != nil {
+		span.Finish()
+	}
+
+	// report all possible spans to the collector
+	tracerCloser.Close()
+}
+
+// trace creates a new tracing span based on the specified contex, subsystem
+// and name.
+func trace(ctx context.Context, subsystem, name string) (opentracing.Span, context.Context) {
```

Comment:
@jodh-intel this is a nice abstraction which almost completely decouples you from the actual underlying tracing implementation. However we return `opentracing.Span` here and then invoke the methods available in that object directly. e.g. finish()

We we provided methods or returned an interface here that we owned, then we could switch tracing frameworks w/o any code change anywhere else.

/cc @sboeuf

Reply:
This will be handled in a follow on PR, correct?  Is there an issue in place already which we can reference, @mcastelino  @sboeuf ?

Reply:
Yep - the plan is I'll pass the baton to @mcastelino to switch the agent to opencensus once this lands.